src/utils.jl

```diff
@@ -54,17 +53,10 @@ function activation(hard, tile)
   buff
 end
 
-array_cpu(array) = array
-
-array_gpu(array) = CuArray{Float32}(array)
-
-const array_kernel = CUDA.functional() ? array_gpu : array_cpu
-
-view_cpu(array, I) = view(array, I)
-
-view_gpu(array, I) = Array(array[I])
-
-const view_kernel = CUDA.functional() ? view_gpu : view_cpu
+abstract type Kernel end
+struct CUDAKernel <: Kernel end
+struct OpenCLKernel <: Kernel end
+struct CPUKernel <: Kernel end
```

Comment:
According to our terminology these are kernel methods, not kernels. Rename to KernelMethod, CUDAMethod, OpenCLMethod, CPUMethod, ...